python-checks/src/main/java/org/sonar/python/checks/XPathCheck.java

```diff
@@ -50,20 +52,32 @@
     defaultValue = "" + DEFAULT_MESSAGE)
   public String message = DEFAULT_MESSAGE;
 
-  @Override
-  public String getXPathQuery() {
-    return xpathQuery;
-  }
+  private AstNodeXPathQuery<Object> query = null;
 
   @Override
-  public String getMessage() {
-    return message;
+  public void init() {
+    if (!Strings.isNullOrEmpty(xpathQuery)) {
+      try {
+        query = AstNodeXPathQuery.create(xpathQuery);
+      } catch (RuntimeException e) {
+        throw new IllegalStateException("Unable to initialize the XPath engine, perhaps because of an invalid query: " + xpathQuery, e);
+      }
+    }
   }
 
   @Override
   public void visitFile(@Nullable AstNode fileNode) {
-    if (fileNode != null) {
-      super.visitFile(fileNode);
+    if (query != null && fileNode != null) {
+      List<Object> objects = query.selectNodes(fileNode);
+
+      for (Object object : objects) {
+        if (object instanceof AstNode) {
+          AstNode astNode = (AstNode) object;
+          addLineIssue(message, astNode.getTokenLine());
+        } else if (object instanceof Boolean && (Boolean) object) {
+          addFileIssue(message);
+        }
```

Comment:
@pynicolas I assume you meant "Can you create an XPath query which would NOT give an AstNode or a Boolean?". Yes I can, and it is actually already there in the unit tests, in XPathCheckTest#integer_xpath_result. The XPath query is formally correct but is not relevant for this rule. The rule silently ignores it, as "object" is of type Integer.